onnxruntime/core/graph/contrib_ops/reverse_sequence_schema_defs.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "reverse_sequence_schema_defs.h"
+#include "core/graph/constants.h"
+#include "core/graph/op.h"
+#include <cmath>
+#include <type_traits>
+
+namespace onnxruntime {
+namespace contrib {
+
+using ::ONNX_NAMESPACE::AttributeProto;
+using ::ONNX_NAMESPACE::OPTIONAL;
+using ::ONNX_NAMESPACE::OpSchema;
+using ::ONNX_NAMESPACE::InferenceContext;
+using ::ONNX_NAMESPACE::TensorShapeProto;
+using ::ONNX_NAMESPACE::TensorProto;
+using ::ONNX_NAMESPACE::TensorProto_DataType;
+
+static const char* ReverseSequence_ver1_doc = R"DOC(
+Reverse batch of sequences having different lengths specified by `sequence_lens`.
+
+For each slice i iterating on batch axis, the operator reverses the first sequence_lens[i] elements on time axis.
+
+Example 1:
+  input = [
+    [0.0, 1.0, 2.0],
+    [3.0, 4.0, 5.0],
+  ]
+  sequence_lens = [3, 3]
+  data_format = "batch_major"
+
+  output = [
+    [2.0, 1.0, 0.0],
+    [5.0, 4.0, 3.0],
+  ]
+
+Example 2:
+  input = [
+    [0.0, 1.0, 2.0],
+    [3.0, 4.0, 5.0],
+  ]
+  sequence_lens = [2, 1]
+  data_format = "batch_major"
+
+  output = [
+    [1.0, 0.0, 2.0],
+    [3.0, 4.0, 5.0],
+  ]
+)DOC";
+
+static const char* Input_Data_Format_ver1_doc = R"DOC(
+(Optional) Specify if the input data format is time major (e.g. [seq_length, batch_size, ...]),
+or batch major (e.g. [batch_size, seq_length, ...]). Must be one of time_major (default), or batch_major.
+)DOC";
+
+OpSchema& RegisterReverseSequenceOpSchema(OpSchema&& op_schema) {
+  return op_schema
+    .SetDomain(kMSDomain)
+    .SinceVersion(1)
+    .TypeConstraint(
+      "T",
+      OpSchema::all_tensor_types(),
+      "Input and output types can be of any tensor type.")
+    .TypeConstraint(
+      "T1", {"tensor(int32)"}, "Constrain seq_lens to integer tensor.")
+    .Attr(
+      "data_format",
+      Input_Data_Format_ver1_doc,
+      AttributeProto::STRING,
+      std::string("time_major"))
+    .Input(
+        0,
+        "input",
+        "Tensor of rank r >= 2, with the shape of `[seq_length, batch_size, ...]` or `[batch_size, seq_length, ...]`",
+        "T")
```

Comment:
I noticed TF.reverse_sequence supports any axis being the sequence or batch axis. This is much simpler and easier to implement, but just wondering if there are any known use cases where the extra flexibility would be required. We can avoid until they come up or get prioritized though.

Reply:
Thanks @skottmckay , this is a good point! 
@linkerzhang for some ideas. definition here fixed the input tensor's format, e.g. assume batch_axis and time_axis in the first two axes. While there is another way doing that. We removed "data_format" attribute, and add other two axis, namely "batch_axis", "time_axis" being INT. Then input can have no constrains on formatting. Do you think we need do that from perspective of creating a ONNX op. I am glad to do both way, feel free to let me know your suggestions. 

Another thing is about a corner case: e.g. do we allow sequence_length contain zero. I think this might also align with other ops like LSTM. If they did not allow, here we should not as well. @linkerzhang

Reply:
The other RNN ops don't explicitly say whether a sequence length of zero is supported. Currently we reject it. 

This is from rnn_helpers.cc and is used by RNN/LSTM/GRU
 
    if (std::any_of(sequence_len_entries.cbegin(),
                    sequence_len_entries.cend(),
                    [seq_length](int len) { return len <= 0 || len > seq_length; })) {
      return ORT_MAKE_STATUS(
          ONNXRUNTIME, INVALID_ARGUMENT,
          "Invalid value/s in sequence_lens. All values must be > 0 and < seq_length. seq_length=", seq_length);
    }


---
In reply to: [268498304](https://github.com/Microsoft/onnxruntime/pull/701#discussion_r268498304) [](ancestors = 268498304)

Reply:
(a) I suggest permitting a sequence length of zero. It is easy to support and well-defined (and sooner or later we will have a use-case where there is a zero-length sequence). (b) For the axes, I suggest going with the TF convention in the spec. It will make translating from TF easier and is more general. If we want to restrict it, I suggest using the batch_axis and time_axis attributes, but say that only (0,1) and (1,0) are supported. This will make it easier to generalize it in the future, if we want to.

Reply:
Both a and b sound reasonable to me.

Reply:
Should the restriction of only supporting 0,1 and 1,0 be done in the operator or at the spec level?

Reply:
I was thinking of mentioning the restriction in the operator documentation part of its spec. (I was also under the assumption that this op was being added to the ONNX spec, at which point some of these fine points become more critical.)

Reply:
I update the spec to reflect b. Thanks for the comments!